internal/uvm/create_wcow.go

```diff
@@ -204,10 +154,138 @@ func CreateWCOW(ctx context.Context, opts *OptionsWCOW) (_ *UtilityVM, err error
 		}
 	}
 
-	uvm.scsiLocations[0][0] = &SCSIMount{
-		vm:       uvm,
-		HostPath: doc.VirtualMachine.Devices.Scsi["0"].Attachments["0"].Path,
-		refCount: 1,
+	return doc, nil
+}
+
+// CreateWCOW creates an HCS compute system representing a utility VM.
+// The HCS Compute system can either be created from scratch or can be cloned from a
+// template.
+//
+// WCOW Notes:
+//   - The scratch is always attached to SCSI 0:0
+//
+func CreateWCOW(ctx context.Context, opts *OptionsWCOW) (_ *UtilityVM, err error) {
+	ctx, span := trace.StartSpan(ctx, "uvm::CreateWCOW")
+	defer span.End()
+	defer func() { oc.SetSpanStatus(span, err) }()
+
+	if opts.ID == "" {
+		g, err := guid.NewV4()
+		if err != nil {
+			return nil, err
+		}
+		opts.ID = g.String()
+	}
+
+	span.AddAttributes(trace.StringAttribute(logfields.UVMID, opts.ID))
+	log.G(ctx).WithField("options", fmt.Sprintf("%+v", opts)).Debug("uvm::CreateWCOW options")
+
+	uvm := &UtilityVM{
+		id:                      opts.ID,
+		owner:                   opts.Owner,
+		operatingSystem:         "windows",
+		scsiControllerCount:     1,
+		vsmbDirShares:           make(map[string]*VSMBShare),
+		vsmbFileShares:          make(map[string]*VSMBShare),
+		vpciDevices:             make(map[string]*VPCIDevice),
+		physicallyBacked:        !opts.AllowOvercommit,
+		devicesPhysicallyBacked: opts.FullyPhysicallyBacked,
+	}
+
+	defer func() {
+		if err != nil {
+			uvm.Close()
+		}
+	}()
+
+	if err := verifyOptions(ctx, opts); err != nil {
+		return nil, errors.Wrap(err, errBadUVMOpts.Error())
+	}
+
+	uvmFolder, err := uvmfolder.LocateUVMFolder(ctx, opts.LayerFolders)
+	if err != nil {
+		return nil, fmt.Errorf("failed to locate utility VM folder from layer folders: %s", err)
+	}
+
+	// TODO: BUGBUG Remove this. @jhowardmsft
+	//       It should be the responsiblity of the caller to do the creation and population.
+	//       - Update runhcs too (vm.go).
+	//       - Remove comment in function header
+	//       - Update tests that rely on this current behaviour.
+	// Create the RW scratch in the top-most layer folder, creating the folder if it doesn't already exist.
+	scratchFolder := opts.LayerFolders[len(opts.LayerFolders)-1]
+
+	// Create the directory if it doesn't exist
+	if _, err := os.Stat(scratchFolder); os.IsNotExist(err) {
+		if err := os.MkdirAll(scratchFolder, 0777); err != nil {
+			return nil, fmt.Errorf("failed to create utility VM scratch folder: %s", err)
+		}
+	}
+
+	doc, err := prepareConfigDoc(ctx, uvm, opts, uvmFolder)
```

Comment:
What happens if the user creates a template UVM using image A, but then the clone UVM uses image B (e.g. if the sandbox image is updated between template and clone operations). It seems to me unless we check for this, there could be a mismatch between the base layer VSMB share and the scratch disk that is used.

Probably we should just save the layer path on template creation, and use that for the clone.

Reply:
PR #840 adds a verification which checks the createOptions of template and clone UVMs and errors out if they don't match.